### PR TITLE
Refactor field flags to use an object an not strpos

### PR DESCRIPTION
**Pull request by williamdes**
Refactor field flags to use an object an not strpos

Ref: #16005

### CODE REVIEW

File: libraries/classes/Sql.php [hunk 1]
```diff
@@ -1273,6 +1275,7 @@ private function getHtmlForSqlQueryResultsTable(
                     $fields_meta = $this->dbi->getFieldsMeta($result);
                     if (! is_array($fields_meta)) {
                         $fields_cnt = 0;
+                        $fields_meta = [];// TODO: evaluate impacts of this hack
```
Comment (williamdes) [inline libraries/classes/Sql.php:1278]: todo

File: templates/table/chart/tbl_chart.twig [hunk 2]
```diff
@@ -69,6 +69,8 @@
                 </label>
                 <select name="chartSeries" id="select_chartSeries" multiple="multiple">
                     {% for idx, key in keys %}
+                      {% set numeric_types = ['int', 'real'] %}
+                        {# FIXME: use getters #}
```
Comment (williamdes) [inline templates/table/chart/tbl_chart.twig:73]: todo

File: libraries/classes/FieldMetadata.php [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    {
+        return $this->mappedType === $type;
+    }
+
+    /**
+     * Check if it is NOT the mapped type
+     *
+     * @phpstan-param self::TYPE_* $type
+     */
+    public function notType(string $type): bool
```
Comment (mauriciofauth) [inline libraries/classes/FieldMetadata.php:369]: This function is redundant, isn't it? If not, I think that `isNotType` is a better name.

Comment (williamdes) [reply libraries/classes/FieldMetadata.php:369]: Yes, it is but it produces clearer code
I will rename it :)

it is quite easy to miss or remove an `!` in particular in multi lines expressions

File: libraries/classes/Display/Results.php [hunk 4]
```diff
@@ -4309,11 +4305,12 @@ private function getSortedColumnMessage(
         // check for non printable sorted row data
         $meta = $fields_meta[$sorted_column_index];
 
-        if (stripos($meta->type, self::BLOB_FIELD) !== false
-            || ($meta->type === self::GEOMETRY_FIELD)
-        ) {
+        $blobOrGeometry = $meta->isType(FieldMetadata::TYPE_BLOB)
```
Comment (mauriciofauth) [inline libraries/classes/Display/Results.php:4308]:
```suggestion
        $isBlobOrGeometry = $meta->isType(FieldMetadata::TYPE_BLOB)
```

**Review by MauricioFauth — CHANGES_REQUESTED** [review]

**Comment by williamdes** [comment]
I applied your comments @mauriciofauth, should be okay now

**Review by MauricioFauth — APPROVED** [review]

File: libraries/classes/FieldMetadata.php [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    public $length;
+
+    /**
+     * A field only used by the Results class
+     *
+     * @var string
+     */
+    public $internalMimeType;
+
+    public function __construct(int $fieldType, int $fieldFlags, stdClass $field)
```
[the remark this answers is not in the document]
Comment (williamdes) [reply libraries/classes/FieldMetadata.php:163]: I will try but I think it is not php7.1 compatible

[the remark this answers is not in the document]
Comment (mauriciofauth) [reply libraries/classes/FieldMetadata.php:163]: Master is PHP 7.2+ now.

File: libraries/classes/FieldMetadata.php [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+
+    /**
+     * @see https://dev.mysql.com/doc/connectors/en/apis-php-mysqli.constants.html
+     */
+    private function getTypeMap(): array
+    {
+        // Issue #16043 - client API mysqlnd seem not to have MYSQLI_TYPE_JSON defined
+        if (! defined('MYSQLI_TYPE_JSON')) {
+            define('MYSQLI_TYPE_JSON', 245);
+        }
```
Comment (mauriciofauth) [inline libraries/classes/FieldMetadata.php:218]: Looks like `MYSQLI_TYPE_JSON` is only defined if the server supports the JSON field type and PHP is using mysqlnd (MySQL 5.7.8+ and MariaDB 10.2.7+, but I didn't test this).

https://github.com/php/php-src/blob/php-8.0.2/ext/mysqli/mysqli.c#L760-L762
https://github.com/php/php-src/blob/php-8.0.2/ext/mysqli/tests/mysqli_constants.phpt#L200-L202
https://mariadb.com/kb/en/json-data-type/
https://dev.mysql.com/doc/refman/5.7/en/json.html

Comment (williamdes) [reply libraries/classes/FieldMetadata.php:218]: Yeah, but It did introduce so much problems I will not take the risk another time to remove it ^^
Just see https://github.com/phpmyadmin/phpmyadmin/issues/16043#issuecomment-616465462 and the issue to find out how narrow this fix is

Comment (mauriciofauth) [reply libraries/classes/FieldMetadata.php:218]: This should not be removed. I think I removed last time based when this constant was introduced in PHP, but I didn't knew back then that the availability of this constant is based on the database server version.

**Comment by williamdes** [comment]
I am working on more improvements, my last finding is that just calling a getter that returns a boolean costs a lot, set the field to public and access it costs nothing.

File: libraries/classes/FieldMetadata.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (williamdes) [reply libraries/classes/FieldMetadata.php]: This had a very funny side effect, I can now feed in a FieldMetadata to create a FieldMetadata. Added a test for that :smile:

**Comment by williamdes** [comment]
### Bench specs

- 1 123 123 rows of the French fantoir database.
- PHP 7.4
- All in Docker instances
- Using SQL export with no compression method
- Dump size: `169.3 MiB (177 563 061 octets)`

### Conclusions

Before my last bunch of commits I used functions (getters) to access booleans, that was costing sooo much time the export would take `106,559,306 microsecs` for example.

### Before the refactoring

The code exports rows and the network rate is ~2.2Mo/s
Total Incl. Wall Time (microsec): 72,292,560 microsecs
Number of Function Calls: 93,280,439

#### Call graph

xhprof graph

### After the refactoring

The code exports rows and the network rate is ~2.8Mo/s to ~3Mo/s

Total Incl. Wall Time (microsec): 55,983,686 microsecs
Number of Function Calls: 71,941,181

#### Call graph

xhprof graph

File: libraries/classes/Display/Results.php [hunk 7]
```diff
@@ -2013,34 +2005,34 @@ private function getSortOrderLink(
...
      */
-    private function getClassForNumericColumnType($fields_meta, array &$th_class)
+    private function getClassForNumericColumnType(FieldMetadata $fields_meta, array &$th_class)
     {
-        if (! preg_match(
-            '@int|decimal|float|double|real|bit|boolean|serial@i',
-            (string) $fields_meta->type
-        )) {
+        // '@int|decimal|float|double|real|bit|boolean|serial@i'
+        //TODO: re-evaluate
```
Comment (mauriciofauth) [inline libraries/classes/Display/Results.php:2016]: This will be done before merge?

Comment (williamdes) [reply libraries/classes/Display/Results.php:2016]: Yup, I am checking why this was added
Ref: https://github.com/phpmyadmin/phpmyadmin/issues/4746#issuecomment-784306563
Ref: https://github.com/phpmyadmin/phpmyadmin/commit/b661cd7c9b31f8bc564d2f9a1b8527e0eb966de8

File: libraries/classes/FieldMetadata.php [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+     * @var int
+     */
+    public $length;
+
+    /**
+     * A field only used by the Results class
+     *
+     * @var string
+     */
+    public $internalMimeType;
```
Comment (mauriciofauth) [inline libraries/classes/FieldMetadata.php:225]:
```suggestion
    public $internalMediaType;
```
See: https://github.com/phpmyadmin/phpmyadmin/issues/15350

**Review by MauricioFauth — CHANGES_REQUESTED** [review]

File: libraries/classes/FieldMetadata.php (hunk 8 above)
Comment (williamdes) [reply libraries/classes/FieldMetadata.php:225]: Good catch, one less occurrence of mime type

**Comment by williamdes** [comment]
Here is a test table for this PR

```sql
-- phpMyAdmin SQL Dump
-- version 5.2.0-dev
-- https://www.phpmyadmin.net/
--
-- Hôte : mariadb104.williamdes.local:3306
-- Généré le : mer. 24 fév. 2021 à 22:30
-- Version du serveur :  10.4.10-MariaDB-1:10.4.10+maria~bionic
-- Version de PHP : 8.0.1

SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
START TRANSACTION;
SET time_zone = "+00:00";

--
-- Base de données : `numbers_align_gauche`
--

-- --------------------------------------------------------

--
-- Structure de la table `t1`
--

CREATE TABLE `t1` (
  `int` int(11) NOT NULL,
  `tinyint` tinyint(4) NOT NULL,
  `smallint` smallint(6) NOT NULL,
  `bigint` bigint(20) NOT NULL,
  `uzint` int(10) UNSIGNED ZEROFILL NOT NULL,
  `uztinyint` tinyint(3) UNSIGNED ZEROFILL NOT NULL,
  `uzsmallint` smallint(5) UNSIGNED ZEROFILL NOT NULL,
  `uzbigint` bigint(20) UNSIGNED ZEROFILL NOT NULL,
  `uint` int(10) UNSIGNED NOT NULL,
  `utinyint` tinyint(3) UNSIGNED NOT NULL,
  `usmallint` smallint(5) UNSIGNED NOT NULL,
  `ubigint` bigint(20) UNSIGNED NOT NULL,
  `real` double NOT NULL,
  `ureal` double UNSIGNED NOT NULL,
  `uzreal` double UNSIGNED ZEROFILL NOT NULL,
  `float` float NOT NULL,
  `ufloat` float UNSIGNED NOT NULL,
  `uzfloat` float UNSIGNED ZEROFILL NOT NULL,
  `double` double NOT NULL,
  `udouble` double UNSIGNED NOT NULL,
  `uzdouble` double UNSIGNED ZEROFILL NOT NULL,
  `decimal` decimal(10,0) NOT NULL,
  `udecimal` decimal(10,0) UNSIGNED NOT NULL,
  `uzdecimal` decimal(10,0) UNSIGNED ZEROFILL NOT NULL,
  `bit` bit(1) NOT NULL,
  `ubit` bit(1) NOT NULL,
  `zbit` bit(1) NOT NULL,
  `boolean` tinyint(1) NOT NULL,
  `uboolean` tinyint(1) UNSIGNED NOT NULL,
  `uzboolean` tinyint(1) UNSIGNED ZEROFILL NOT NULL,
  `serial` bigint(20) UNSIGNED NOT NULL,
  `varchar` varchar(11) NOT NULL,
  `date` date NOT NULL,
  `datetime` datetime NOT NULL,
  `time` time NOT NULL,
  `year` year(4) NOT NULL,
  `tinyblob` tinyblob NOT NULL,
  `binary` binary(11) NOT NULL,
  `bvarchar` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  `char` char(4) NOT NULL,
  `bchar` char(4) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  `point` point NOT NULL,
  `blob` blob NOT NULL,
  `mediumblob` mediumblob NOT NULL,
  `longblob` longblob NOT NULL,
  `btext` text CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  `text` text NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

--
-- Déchargement des données de la table `t1`
--

INSERT INTO `t1` (`int`, `tinyint`, `smallint`, `bigint`, `uzint`, `uztinyint`, `uzsmallint`, `uzbigint`, `uint`, `utinyint`, `usmallint`, `ubigint`, `real`, `ureal`, `uzreal`, `float`, `ufloat`, `uzfloat`, `double`, `udouble`, `uzdouble`, `decimal`, `udecimal`, `uzdecimal`, `bit`, `ubit`, `zbit`, `boolean`, `uboolean`, `uzboolean`, `serial`, `varchar`, `date`, `datetime`, `time`, `year`, `tinyblob`, `binary`, `bvarchar`, `char`, `bchar`, `point`, `blob`, `mediumblob`, `longblob`, `btext`, `text`) VALUES
(1, 1, 1, 1, 0000000001, 001, 00001, 00000000000000000001, 1, 1, 1, 1, 1.2, 1.2, 00000000000000000001.2, 1.2, 1.2, 0000000001.2, 1.2, 1.2, 00000000000000000001.2, '1', '1', '0000000001', b'1', b'1', b'1', 1, 1, 1, 1, '1.2', '2021-01-01', '2021-02-24 15:18:54', '00:00:01', 2021, 0x03c3f7068700a2f2a2a0a202a204d696e696d616c20636f6e6669672066696c6520666f7220746573740a202a2f0a0a6465636c617265287374726963745f74797065733d3, 0x1111000000000000000000, '123\r\n123', '1111', '111', 0x00000000010100000000000000000000000000000000000000, 0x3c3f7068700a2f2a2a0a202a204d696e696d616c20636f6e6669672066696c6520666f7220746573740a202a2f0a0a6465636c617265287374726963745f74797065733d31293b0a0a2469203d20303b0a0a2f2a205365727665723a206c6f63616c686f7374205b315d202a2f0a24692b2b3b0a246366675b2753657276657273275d5b24695d5b27766572626f7365275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27686f7374275d203d20276c6f63616c686f7374273b0a246366675b2753657276657273275d5b24695d5b27706f7274275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27736f636b6574275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27617574685f74797065275d203d2027636f6f6b6965273b0a, 0x3c3f7068700a2f2a2a0a202a204d696e696d616c20636f6e6669672066696c6520666f7220746573740a202a2f0a0a6465636c617265287374726963745f74797065733d31293b0a0a2469203d20303b0a0a2f2a205365727665723a206c6f63616c686f7374205b315d202a2f0a24692b2b3b0a246366675b2753657276657273275d5b24695d5b27766572626f7365275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27686f7374275d203d20276c6f63616c686f7374273b0a246366675b2753657276657273275d5b24695d5b27706f7274275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27736f636b6574275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27617574685f74797065275d203d2027636f6f6b6965273b0a, 0x3c3f7068700a2f2a2a0a202a204d696e696d616c20636f6e6669672066696c6520666f7220746573740a202a2f0a0a6465636c617265287374726963745f74797065733d31293b0a0a2469203d20303b0a0a2f2a205365727665723a206c6f63616c686f7374205b315d202a2f0a24692b2b3b0a246366675b2753657276657273275d5b24695d5b27766572626f7365275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27686f7374275d203d20276c6f63616c686f7374273b0a246366675b2753657276657273275d5b24695d5b27706f7274275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27736f636b6574275d203d2027273b0a246366675b2753657276657273275d5b24695d5b27617574685f74797065275d203d2027636f6f6b6965273b0a, '123\r\n123', '123\r\n123'),
(123, 123, 123, 123, 0000000123, 123, 00123, 00000000000000000123, 123, 123, 123, 123, 123.1, 123.1, 00000000000000000123.1, 123.1, 123.1, 0000000123.1, 123.1, 123.1, 00000000000000000123.1, '123', '123', '0000000123', b'1', b'0', b'1', 123, 123, 123, 123, '', '0000-00-00', '0000-00-00 00:00:00', '00:00:00', 0000, '', 0x0000000000000000000000, '', '', '', 0x, '', '', '', '', '');

--
-- Index pour les tables déchargées
--

--
-- Index pour la table `t1`
--
ALTER TABLE `t1`
  ADD UNIQUE KEY `serial` (`serial`),
  ADD UNIQUE KEY `float` (`float`),
  ADD UNIQUE KEY `ufloat` (`ufloat`);

--
-- AUTO_INCREMENT pour les tables déchargées
--

--
-- AUTO_INCREMENT pour la table `t1`
--
ALTER TABLE `t1`
  MODIFY `serial` bigint(20) UNSIGNED NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=124;

--
-- Contraintes pour les tables déchargées
--

--
-- Contraintes pour la table `t1`
--
ALTER TABLE `t1`
  ADD CONSTRAINT `ff` FOREIGN KEY (`float`) REFERENCES `t2` (`float`),
  ADD CONSTRAINT `uf` FOREIGN KEY (`ufloat`) REFERENCES `t2` (`ufloat`);
COMMIT;
```

**Comment by williamdes** [comment]
I did not detect any notable difference on table Results output, It is now ready to merge after approval.

**Review by MauricioFauth — APPROVED** [review]
Really good work. :100:

**Pull request merged**